### DISCUSSION TITLE
`--type` (or `--type-add`) for files with no extension?

### DISCUSSION THREAD

**Discussion by thomcc** [BurntSushi/ripgrep#1967] [Q&A] [answered]
I'd like to search a folder for some C++ code. It also has a great deal of other code, so I'd like to limit it to just the C++, as you'd normally do with -tcpp. Unfortunately, some<sup>1</sup> of these files are in a pseudo c++-stdlib, and the convention for the C++ standard library is for the contents to have no file extension<sup>2</sup>.

So, I'd like to do something like `ripgrep -tcpp -tnoext $path -e $pat`, where `noext` matches files without a file extension. 

It feels like it would be nice if something for this existed in the builtin list (probably with a better name than `noext`), but honestly it's fine if I have to add the `noext` type via `--type-add` somehow... Sadly, I couldn't figure out how to express "no file extension" (or more generally files without `.` in their names) in a glob.

I think there's a reasonable chance this is possible with the glob, but my glob-fu was just too weak. It's also quite possibly doable through other means I didn't think of. (And if neither of those is true, then maybe something should be added? It's an edge case, but not an inconceivable one)

---

1: In fact, plausibly the ones holding the data I'm looking for, although I'm not sure. (I have to check several branches, which is why I'd like a better method here).

2: In case you doubt this, see most of the files inside https://github.com/llvm/llvm-project/tree/main/libcxx/include.

**Comment by BurntSushi** [comment, marked as answer]
So the first thing I did was find a query that gives me a hit in a `.cpp` file, a hit in a cpp file with no extension and a hit in a non-cpp file:

```
$ rg 'ABI compatible manner|Splice the instructions of the|to learn about the layout of the source code tree'
README.md
106:to learn about the layout of the source code tree.

libcxx/include/optional
362:// make the change in an ABI compatible manner.

mlir/lib/Transforms/Utils/InliningUtils.cpp
236:  // Splice the instructions of the inlined entry block into the insert block.
```

So with that in our back pocket, I don't think there is any way to express "match a file name without an extension" in the glob syntax supported by ripgrep. The problem is that standard globbing (leaving out "extended glob" syntax) is fundamentally limited. The only repetition operators are `*` and `**`, and neither provide any control over _what_ is being repeated. `*` for example just means "any character zero or more times." Therefore, the only real way to search only files with no extension is to write a glob that _matches_ any file path with a `.` and then exclude files matching that pattern:

```
$ rg 'ABI compatible manner|Splice the instructions of the|to learn about the layout of the source code tree' -g '!*.*'
libcxx/include/optional
362:// make the change in an ABI compatible manner.
```

So you might think you can compose this with `-g '*.cpp'` to get what you want...

```
$ rg 'ABI compatible manner|Splice the instructions of the|to learn about the layout of the source code tree' -g '!*.*' -g '*.cpp'
mlir/lib/Transforms/Utils/InliningUtils.cpp
236:  // Splice the instructions of the inlined entry block into the insert block.
```

But this doesn't quite work because once a whitelist rule is introduced (that's the `*.cpp` glob), that introduces a new requirement that a file path _must_ match at least one of those whitelist rules in order to be searched. Otherwise, whitelist rules would just not work how you might expect. Since `*.cpp` doesn't match files without extensions, those paths aren't considered.

However, globs are applied in precedence order. And since `!*.*` excludes everything with a `.` in it, we can use that to our advantage by matching "everything" at the lowest precedence:

```
$ rg 'ABI compatible manner|Splice the instructions of the|to learn about the layout of the source code tree' -g '*' -g '!*.*' -g '*.cpp'
libcxx/include/optional
362:// make the change in an ABI compatible manner.

mlir/lib/Transforms/Utils/InliningUtils.cpp
236:  // Splice the instructions of the inlined entry block into the insert block.
```

So this is in effect saying, "if a path has a cpp extension, search it. Otherwise, if a path has any `.` in it, skip it. Finally, for anything making it to this point, search it." The first rule prevents the second rule from excluding `.cpp` files and the second rule prevents the third rule from matching literally everything.

**Reply by thomcc** [reply]
This helped a great deal, and also helps me feel a little better about the fact that I couldn't figure out the glob that would match this on my own 😅.

I do think it might be worth simplifying the "no file extension" case somehow, but perhaps this thread good enough to serve as documentation for someone in the future.

Thanks a great deal, as always.
